Feature/reports microservice

the-man-w-laughs opened the pull request:
Third microservise that is responsible for creating, deleting and storing reports.

Tech stack:
n-tier;
minio for storing reports;
mongo for storing reports info;
http for communication;
automapper.

the-man-w-laughs committed:
Add initial structure for ReportHub.

the-man-w-laughs committed:
Implement GetAllProjectById.

the-man-w-laughs committed:
Add minio storage.

the-man-w-laughs committed:
Refactor ReportHub service.

the-man-w-laughs committed:
Refactor ReportHub service.

the-man-w-laughs committed:
Add gateway project.

the-man-w-laughs committed:
Add routes to ocelot.json

the-man-w-laughs committed:
Add pagination to ReportInfoController.

the-man-w-laughs committed:
Add configuration for docker

Major changes:
- Fix error in ReportHub,
now to make http request to TeamHub ReportHub includes only Authorizaiton headers;
- add dockerfile for apigateway;
- add apigateway to docker-compose.

katyalys commented on src/Shared/Middleware/ExceptionMiddleware.cs:
```diff
@@ -0,0 +1,81 @@
+
+    private async Task HandleExceptionAsync(HttpContext context, Exception exception)
+    {
+        context.Response.ContentType = "application/json";
```
empty line before switch

katyalys commented on src/Shared/Middleware/ExceptionMiddleware.cs:
```diff
@@ -0,0 +1,81 @@
+    public async Task InvokeAsync(HttpContext httpContext)
+    {
+        await _next(httpContext);
+        try { }
```
empty line before try

katyalys commented on src/Shared/Repository/Sql/IRepository.cs:
```diff
@@ -0,0 +1,18 @@
+    Task<List<TEntity>> GetAllAsync(Expression<Func<TEntity, bool>> where);
+    Task<TEntity?> GetByIdAsync(int id);
+    Task<TEntity?> GetAsync(Expression<Func<TEntity, bool>> where);
+    Task<TEntity> AddAsync(TEntity entity);
```
add cancellation token as parameter to async methods

the-man-w-laughs committed:
Merge branch 'feature/reports-microservice' into feature/gateway

the-man-w-laughs committed:
Merge pull request #4 from the-man-w-laughs/feature/gateway

Feature/gateway

the-man-w-laughs committed:
Add cancellation token (TeamHub).

Major changes:
- add cancellation token to every async operation in Repository (base-repository);
- add cancellation token to every handler inside TeamHub.WebApi controllers;
- pass cancellation token in repositories methods invocations.

the-man-w-laughs committed:
Fix exception middleware.

the-man-w-laughs committed:
Fix exception middleware.

the-man-w-laughs committed:
Add message publisher to Identity.

Major changes:
- add UserCreatedMessage;
- add UserDeletedMessage.

the-man-w-laughs committed:
Add consumer to TeamHub.

Major changes:
- add create user handler;
- add delete user handler.

the-man-w-laughs committed:
Add consumer to ReportHub.

Major changes:
- add handler on users delete (all users projects reports will be deleted).

the-man-w-laughs committed:
Add repositories for Redis

Major changes:
- Add RedisRepository (base repository);
- Add UserRequestRepository (concrete implementation to store time of user's latest reuquest);

the-man-w-laughs committed:
Add cacher to ApiGateway;

Major changes:
- register IUserRequestRepository (to access redis);
- add UserChageMiddleware (to set last user's request).

the-man-w-laughs committed:
Add redis to TeamHub.

Major changes:
- add extra field to UserResponseDto (IsOnline);
- put additional logic to user automapper to make request to the redis to set isOnline field.

the-man-w-laughs committed:
Refactor ApiGateway

the-man-w-laughs committed:
Remove user caching from api gateway.

the-man-w-laughs committed:
Final implementation of redis caching

Major changes:
- add IsOnline property to UserDtos;
- add value resolvers for UserDtos in identity and teamhub (to set IsOnline property);
- add middleware to cache time of the latest user's request.

Description:
First I tried to put caching to api gateway but It didn't work at all, so I decided to put this middleware in each microservice.

the-man-w-laughs committed:
Refactor RabbitMQ configuration

Major changes:
- add RabbitMQSettings;
- rewrite MassTransit registration using IOptions;
- optimise multiple file deletion;
- make namespaces names start with capital letter.

the-man-w-laughs committed:
Merge pull request #9 from the-man-w-laughs/feature/redis

Feature/redis

the-man-w-laughs committed:
Merge branch 'feature/add-teamhub-microservice' into feature/reports-microservice

the-man-w-laughs committed:
Make controller name plural.

the-man-w-laughs committed:
Refactor code

Major changes
- add async suffixes to all async methods.

the-man-w-laughs committed:
Merge branch 'feature/reports-microservice' into feature/rabbitMQ

the-man-w-laughs committed:
Merge pull request #8 from the-man-w-laughs/feature/rabbitMQ

Feature/rabbit mq
